Add additional resource types and filtering by resource subtype

ronitjadhav opened the pull request:
## Overview
This PR improves the QGIS Hub Plugin's ability to handle resource types by making it more dynamic and future-proof. It also adds support for displaying model dependencies information and filtering by resource subtypes.

## Key changes

### Dynamic resource type handling
- Added support for the 'Map' resource type that was recently added to the QGIS Hub API
- Implemented a system to dynamically discover, register, and display new resource types without requiring code changes
- The plugin will now automatically create categories for any new resource types that are added to the QGIS Hub API in the future

### Resource subtype filtering
- Added support for filtering resources by their subtype
- Resources are now organized in a tree hierarchy with resource types and their subtypes
- Improved resource type tree display with proper sizing to accommodate long subtype names
- Added automatic expansion of categories to make subtypes immediately visible

### Model dependencies support
- Added UI elements to show the dependencies for model resources
- Dependencies are displayed in the preview panel when a model with dependencies is selected
- Dependencies section only appears when relevant (for models that actually have dependencies)

### Close Button
- Added a close button to close the resource preview window

### Renamed Resource Type Labels
- Renamed the Resource Type labels to match with the QGIS Hub Website

### UI Improvements
- Enhanced overall interface with better padding and spacing between elements
- Added proper sizing for buttons and input controls for better usability
- Improved the preview panel layout with consistent form spacing
- Made the icon size slider only visible in the icon view mode
- Fixed various layout issues to create a more polished and professional UI

### Note

- Processing Scripts are still not available in the API, so it will be added automatically when it is available in the future. (https://hub.qgis.org/api/v1/resources/?resource_type=processingscript) 


### Screenshots

![image](https://github.com/user-attachments/assets/561684eb-e223-48ec-bc66-02eb8c995c90)

![image](https://github.com/user-attachments/assets/7c03b4f9-e7ab-4969-9e6b-17f09759bc61)


This enhancement makes the plugin more maintainable by eliminating the need for code updates when new resource types are added to the QGIS Hub API.

#106

ismailsunni reviewed (approved):
LGTM

ismailsunni commented:
Some improvements:
- Put the close button for the preview somewhere else, currently it takes a quite big space
- For some reason, the thumbnail in the preview is not centered. Could you fix it, only for some resource as seen below?
   ![image](https://github.com/user-attachments/assets/33864bc6-e9fa-409b-ba57-c928c906b545)
- For Processing Script, we can add a button to add it directly to QGIS directory. See https://www.qgistutorials.com/en/docs/processing_python_scripts.html for reference (might help)
- To make it more uniform, perhaps we can also show the number of resources in each resource type as you did in the resource subtype
- Lastly, great job. Please put some changelog in CHANGELOG.md for version 0.4.0

Thanks!

ronitjadhav commented:
@ismailsunni 
Thanks for the review.

1. Any suggestion where I can put the close button? or should I remove it completely for now?
2. I am not able to reproduce this issue, for me all the thumbnails are centered.
3. Currently we don't have any processing scripts available in the API to implement this (Maybe we can do it in another PR)
4. Implemented
5. Implemented

ismailsunni commented:
1. Let's discuss on our meeting today
2. Strange, I checked it again, I couldn't reproduce it
3. Let's discuss on our meeting today, there is one new processing script currently

ronitjadhav commented:
I have added a button to add the processing script to QGIS.
But the current processing script has some issue so it does not successfully add it to the processing toolbox.

` CRITICAL    Could not import script algorithm 'pcraster_hand' from '/home/rjadhav/.local/share/QGIS/QGIS3/profiles/default/processing/scripts/qgis_hub/pcraster_hand.py'
             No module named 'pcraster'`

jvdkwast commented:
> @ismailsunni  I have added a button to add the processing script to QGIS.
> But the current processing script has some issue so it does not successfully add it to the processing toolbox.
> 
> We can see the following log message:
> 
> ` CRITICAL    Could not import script algorithm 'pcraster_hand' from '/home/rjadhav/.local/share/QGIS/QGIS3/profiles/default/processing/scripts/qgis_hub/pcraster_hand.py'
>              No module named 'pcraster'`

The script has a dependency (in metadata). When a script is installed, the packages are imported and in this case the error is related to missing pcraster. I think this behaviour is good, so the user needs to first install the dependency (plugin and/or Python package) before installing the script.

ismailsunni commented:
thanks @jvdkwast 
I agree with you. We can also add a better error message when failed installing the script. So that the user know what to do (i.e. installing the dependencies).
